File: Auth/Users.App/ResetPasswordService.cs
using Microsoft.Extensions.Options;
using System.Security.Cryptography;
using Users.App.Interface;
using Users.App.Settings;

namespace Users.App
{
    public class ResetPasswordService(
        IUserRepo repo,
        IResetPasswordEmailService emailService,
        IOptions<ResetPasswordSettings> resetPasswordSettings)
    {
        private readonly ResetPasswordSettings _resetPasswordSettings = resetPasswordSettings.Value;

        public async Task<bool> SendResetPasswordEmail(string email)
        {
            var existingUser = await repo.FindByEmailAsync(email);
            if (existingUser == null || existingUser.EmailConfirmed == false || Constants.IsLocalAccount(existingUser.Issuer) == false)
            {
                return false;
            }

            var resetToken = GenerateResetPasswordToken();
            var resetUrl = $"{_resetPasswordSettings.ResetPasswordUrl}?token={Uri.EscapeDataString(resetToken)}";

            bool sent = await emailService.SendResetPasswordEmailAsync(existingUser, email, resetUrl);
            if (!sent)
            {
                return false;
            }

            var expiresAt = DateTimeOffset.UtcNow.AddMinutes(_resetPasswordSettings.ResetPasswordTokenExpiresInMinutes);
            existingUser.ResetPasswordToken = resetToken;
            existingUser.ResetPasswordTokenExpiresAt = expiresAt;
            await repo.UpdateAsync(existingUser);
            return true;
        }

        public async Task<ResetPasswordResult> ResetPasswordAsync(string token, string newPassword)
        {
            if (string.IsNullOrWhiteSpace(token))
            {
                return new ResetPasswordResult
                {
                    Success = false,
                    Error = "InvalidToken",
                    ErrorDescription = "Token is required."
                };
            }

            var user = await repo.FindByResetPasswordTokenAsync(token);
            if (user == null || Constants.IsLocalAccount(user.Issuer) == false)
            {
                return new ResetPasswordResult
                {
                    Success = false,
                    Error = "InvalidToken",
                    ErrorDescription = "Invalid reset token."
                };
            }

            if (user.ResetPasswordTokenExpiresAt == null ||
                DateTimeOffset.UtcNow > user.ResetPasswordTokenExpiresAt.Value)
            {
                return new ResetPasswordResult
                {
                    Success = false,
                    Error = "TokenExpired",
                    ErrorDescription = "Reset token has expired."
                };
            }

            user.Password = PasswordHasher.Hash(newPassword);
            user.ResetPasswordToken = null;
            user.ResetPasswordTokenExpiresAt = null;
            await repo.UpdateAsync(user);

            return new ResetPasswordResult { Success = true };
        }

        private static string GenerateResetPasswordToken()
        {
            var bytes = new byte[32];
            using var rng = RandomNumberGenerator.Create();
            rng.GetBytes(bytes);
            return Convert.ToBase64String(bytes);
        }
    }

    public class ResetPasswordResult
    {
        public bool Success { get; set; }
        public string? Error { get; set; }
        public string? ErrorDescription { get; set; }
    }
}

File: NodeAuth/src/services/ResetPasswordService.ts
import { randomBytes } from "crypto";

import { ResetPasswordSettings } from "../config/resetPassword";
import { IUserRepo } from "../repos/interfaces/IUserRepo";
import { isLocalAccount } from "../utils/constants";
import { PasswordHasher } from "../utils/passwordHasher";
import { IResetPasswordEmailService } from "./interfaces/IResetPasswordEmailService";

export class ResetPasswordService {
  constructor(
    private readonly repo: IUserRepo,
    private readonly emailService: IResetPasswordEmailService,
    private readonly settings: ResetPasswordSettings,
  ) {}

  async sendResetPasswordEmail(email: string): Promise<boolean> {
    const existingUser = await this.repo.findByEmailAsync(email);
    if (
      !existingUser ||
      !existingUser.emailConfirmed ||
      !isLocalAccount(existingUser.issuer)
    ) {
      return false;
    }

    const resetToken = this.generateResetPasswordToken();
    const resetUrl = `${this.settings.resetPasswordUrl}?token=${encodeURIComponent(
      resetToken,
    )}`;

    const sent = await this.emailService.sendResetPasswordEmailAsync(
      existingUser,
      email,
      resetUrl,
    );
    if (!sent) {
      return false;
    }

    existingUser.resetPasswordToken = resetToken;
    existingUser.resetPasswordTokenExpiresAt = this.addMinutes(
      new Date(),
      this.settings.resetPasswordTokenExpiresInMinutes,
    );
    await this.repo.updateAsync(existingUser);
    return true;
  }

  async resetPasswordAsync(
    token: string,
    newPassword: string,
  ): Promise<ResetPasswordResult> {
    if (!token?.trim()) {
      return {
        success: false,
        error: "InvalidToken",
        errorDescription: "Token is required.",
      };
    }

    const user = await this.repo.findByResetPasswordTokenAsync(token);
    if (!user || !isLocalAccount(user.issuer)) {
      return {
        success: false,
        error: "InvalidToken",
        errorDescription: "Invalid reset token.",
      };
    }

    if (
      !user.resetPasswordTokenExpiresAt ||
      Date.now() > user.resetPasswordTokenExpiresAt.getTime()
    ) {
      return {
        success: false,
        error: "TokenExpired",
        errorDescription: "Reset token has expired.",
      };
    }

    user.password = PasswordHasher.hash(newPassword);
    user.resetPasswordToken = undefined;
    user.resetPasswordTokenExpiresAt = undefined;
    await this.repo.updateAsync(user);

    return { success: true };
  }

  private generateResetPasswordToken(): string {
    return randomBytes(32).toString("base64");
  }

  private addMinutes(date: Date, minutes: number): Date {
    return new Date(date.getTime() + minutes * 60 * 1000);
  }
}

export interface ResetPasswordResult {
  success: boolean;
  error?: string;
  errorDescription?: string;
}
